### PR TITLE
Properly deal with absolute paths in unit tests, part 2

### PR DESCRIPTION
**Pull request by rmohr**
**What this PR does / why we need it**:

This PR removes absolute path dependencies in unit-tests for pod-network setup code. The tricky part in this PR is that the network-info cache stores are deeply hidden inside the setup code. To overcome that without rewriting  too much surroundings, two stores were implemented: `PodInterfaceCacheStore` and `DomainInterfaceStore`.  These two are now injected via a store factory which is passed into the Plugging phases of the networking code. By replacing this factory with a fake factory with in-memory fake-stores, all networking code can be tested without persisting files at absolute privileged paths.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

This PR does not refactor any code into a perfect form. It solely deals with absolute paths to allow proper unit-test execution.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/virt-handler/vm.go [hunk 1]
```diff
@@ -47,6 +47,8 @@ import (
 	"k8s.io/client-go/tools/record"
 	"k8s.io/client-go/util/workqueue"
 
+	cache2 "kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/network/cache"
```
Comment (qinqon) [inline pkg/virt-handler/vm.go:50]: netcache or networkcache feels better.

**Review by qinqon — CHANGES_REQUESTED** [review]
Looks good overall, nothing important.

**Comment by maiqueb** [comment]
Nice improvement; what would be the correct way to extend this cache factory mechanism to also encompass the [VIF caches](https://github.com/kubevirt/kubevirt/blob/2122c2beb050399c79a290f880c3fac7295a18ea/pkg/virt-launcher/virtwrap/network/podinterface.go#L63) used for DHCP ? 

Add a new store type, and have `CacheForPID` return both `DomainInterfaceStore` and this new store type?..

File: pkg/virt-launcher/virtwrap/network/podinterface.go [hunk 2]
```diff
@@ -83,8 +86,8 @@ func writeVifFile(buf []byte, pid, name string) error {
 	return nil
 }
 
-func setPodInterfaceCache(iface *v1.Interface, podInterfaceName string, uid string) error {
-	cache := PodCacheInterface{Iface: iface}
+func setPodInterfaceCache(iface *v1.Interface, podInterfaceName string, vmi *v1.VirtualMachineInstance, cacheFactory cache.InterfaceCacheFactory) error {
+	cache := &cache.PodCacheInterface{Iface: iface}
```
Comment (maiqueb) [inline pkg/virt-launcher/virtwrap/network/podinterface.go:90]: You're shadowing an imported pkg name here.

File: pkg/virt-launcher/virtwrap/network/infocache.go [hunk 3]
```diff
@@ -34,24 +32,87 @@ import (
...
 var virtLauncherCachedPattern = "/proc/%s/root/var/run/kubevirt-private/interface-cache-%s.json"
 
-func CreateVirtHandlerNetworkInfoCache() error {
-	return os.MkdirAll(networkInfoDir, 0755)
+type InterfaceCacheFactory interface {
+	CacheForVMI(vmi *v1.VirtualMachineInstance) PodInterfaceCacheStore
+	CacheForPID(pid string) DomainInterfaceStore
+}
+
+func NewInterfaceCacheFactory() InterfaceCacheFactory {
```
[the remark this answers is not in the document]
Comment (rmohr) [reply pkg/virt-launcher/virtwrap/network/infocache.go]: Done.

File: pkg/virt-launcher/virtwrap/network/network_test.go [hunk 4]
```diff
@@ -26,43 +26,49 @@ import (
...
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/network/cache/fake"
+
 	v1 "kubevirt.io/client-go/api/v1"
 )
 
 var _ = Describe("Network", func() {
 	var mockpodNIC *MockpodNIC
 	var ctrl *gomock.Controller
 	var pid int
+	var cacheFactory cache.InterfaceCacheFactory
```
[the remark this answers is not in the document]
Comment (rmohr) [reply pkg/virt-launcher/virtwrap/network/network_test.go:40]: Are you sure? the `fake` cache is not public. Therefore I think I have to take the interface.

File: pkg/virt-launcher/virtwrap/network/podinterface.go (hunk 2 above)
Comment (rmohr) [reply pkg/virt-launcher/virtwrap/network/podinterface.go]: Fixed.

File: pkg/virt-handler/vm.go (hunk 1 above)
Comment (rmohr) [reply pkg/virt-handler/vm.go]: Done.

File: pkg/virt-launcher/virtwrap/network/fake/infocache.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+		domainCacheStores: map[string]cache.DomainInterfaceStore{},
+		lock:              &sync.Mutex{},
+	}
+}
+
+type fakeInterfaceCacheFactory struct {
+	vmiCacheStores    map[types.UID]cache.PodInterfaceCacheStore
+	domainCacheStores map[string]cache.DomainInterfaceStore
+	lock              *sync.Mutex
+}
```
[the remark this answers is not in the document]
Comment (rmohr) [reply pkg/virt-launcher/virtwrap/network/fake/infocache.go]: Done.

**Comment by rmohr** [comment]
Should be ready for another review round.

**Review by maiqueb — COMMENTED** [review]
Just some nits.

You can disregard the comments about commit ordering; only after reviewing the entire thing commit by commit did I realize you were not re-writing past commits.

File: pkg/virt-launcher/virtwrap/network/infocache.go [hunk 6]
```diff
@@ -34,24 +32,87 @@ import (
...
+	Remove() error
+}
+
+type domainInterfaceStore struct {
+	pid string
 }
 
-func CreateVirtHandlerCacheDir(vmiuid types.UID) error {
-	return os.MkdirAll(filepath.Join(networkInfoDir, string(vmiuid)), 0755)
+func (d domainInterfaceStore) Read(iface string) (file *api.Interface, err error) {
```
[the remark this answers is not in the document]
Comment (rmohr) [reply pkg/virt-launcher/virtwrap/network/infocache.go]: Done. I don't mind not using them here.

File: pkg/virt-launcher/virtwrap/network/infocache.go [hunk 7]
```diff
@@ -80,23 +141,29 @@ func readFromCachedFile(obj interface{}, fileName string) error {
...
 }
 
-func readFromVirtLauncherCachedFile(obj interface{}, pid, ifaceName string) error {
+func readFromVirtLauncherCachedFil(obj interface{}, pid, ifaceName string) error {
 	fileName := getInterfaceCacheFile(virtLauncherCachedPattern, pid, ifaceName)
 	return readFromCachedFile(obj, fileName)
 }
 
-func writeToVirtLauncherCachedFile(obj interface{}, pid, ifaceName string) error {
+func writeToVirtLauncherCachedFil(obj interface{}, pid, ifaceName string) error {
```
[the remark this answers is not in the document]
Comment (rmohr) [reply pkg/virt-launcher/virtwrap/network/infocache.go]: Fixed.

File: pkg/virt-launcher/virtwrap/network/infocache.go [hunk 8]
```diff
@@ -34,24 +32,87 @@ import (
...
+func NewDomainInterfaceStore(pid string) DomainInterfaceStore {
+	return domainInterfaceStore{pid: pid}
+}
+
+type podInterfaceCacheStore struct {
+	vmi *v1.VirtualMachineInstance
+}
+
+func (p podInterfaceCacheStore) Read(iface string) (file *PodCacheInterface, err error) {
+	file = &PodCacheInterface{}
```
[the remark this answers is not in the document]
Comment (rmohr) [reply pkg/virt-launcher/virtwrap/network/infocache.go]: Done.

File: pkg/virt-launcher/virtwrap/network/infocache.go [hunk 9]
```diff
@@ -80,23 +141,29 @@ func readFromCachedFile(obj interface{}, fileName string) error {
 	return nil
 }
 
-func readFromVirtLauncherCachedFile(obj interface{}, pid, ifaceName string) error {
+func readFromVirtLauncherCachedFil(obj interface{}, pid, ifaceName string) error {
```
[the remark this answers is not in the document]
Comment (rmohr) [reply pkg/virt-launcher/virtwrap/network/infocache.go]: Fixed.

File: pkg/virt-launcher/virtwrap/network/podinterface.go [hunk 10]
```diff
@@ -30,9 +30,10 @@ import (
 	"strconv"
 	"strings"
 
-	"k8s.io/apimachinery/pkg/types"
 	netutils "k8s.io/utils/net"
 
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/network/cache"
+
```
[the remark this answers is not in the document]
Comment (rmohr) [reply pkg/virt-launcher/virtwrap/network/podinterface.go]: Reformated.

File: pkg/virt-launcher/virtwrap/network/cache/types.go [hunk 11]
```diff
@@ -0,0 +1,9 @@
+package cache
+
+import v1 "kubevirt.io/client-go/api/v1"
+
+type PodCacheInterface struct {
```
[the remark this answers is not in the document]
Comment (rmohr) [reply pkg/virt-launcher/virtwrap/network/cache/types.go:5]: I can. I think I removed it when I started using the new code in follow up commits.

**Comment by rmohr** [comment]
Feedback should be addressed. Thanks.

File: pkg/virt-launcher/virtwrap/network/network_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/virt-launcher/virtwrap/network/network_test.go:40]: fake cache is not public but this test is withing the same package so we can use the concrete time.

[the remark this answers is not in the document]
Comment (rmohr) [reply pkg/virt-launcher/virtwrap/network/network_test.go:40]: Are you really sure? `fake` is a subpackage of `network` as far as I can see.

[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/virt-launcher/virtwrap/network/network_test.go:40]: They are not at the same package so we are all good.

**Review by qinqon — APPROVED** [review]

**Review by davidvossel — APPROVED** [review]
/approve

**Pull request merged**